Core/Src/main_task.c

```diff
@@ -6,33 +6,58 @@
  */
 
 #include "main_task.h"
+#include "global_variables.h"
+#include "debugleds.h"
+#include "main.h"
+
 
 void main_task(void *argument) {
+
 	for(;;) {
 
 		#ifdef DEBUG_ECU
 		extern void brkpt();
 		brkpt();
 		#endif
 
-		//espera receber flag q o botão de RTD foi pressionado
-		osThreadFlagsWait(RTD_BTN_PRESSED_FLAG, osFlagsWaitAny, osWaitForever);
+		osSemaphoreRelease(s_Allowed_change_modeHandle); 							//libera semáforo que permite a mudança de modos
+
+		for(;;) {
+			osThreadFlagsWait(RTD_BTN_PRESSED_FLAG, osFlagsWaitAny, osWaitForever); //espera receber flag q o botão de RTD foi pressionado
+			uint32_t error_flags = osEventFlagsGet(ECU_control_event_id);			//obtem todas as flags
+			error_flags = error_flags & ALL_SEVERE_ERROR_FLAG; 						//filtra apenas flags de erros severos, ignorando as outras
+			if(brake_status && !apps_status && !error_flags && (g_race_mode != ERRO))
+				break;																//sai do for infinito caso tudo esteja certo para acionar RTD
+			else
+				set_debugleds(DEBUGLED1, BLINK, 2);									//envia uma mensagem de alerta caso n seja possível acionar RTD
+		}
 
 		//seta a flag de RTD
-		osEventFlagsSet(ECU_control_event_id, RTD_FLAG);
+		osEventFlagsSet(ECU_control_event_id, RTD_FLAG);					//Seta flag de RTD
+		osSemaphoreAcquire(s_Allowed_change_modeHandle, osWaitForever);		//Bloqueia mudança de modo
+		aciona_sirene();
+
 
 		//espera por qualquer erro relatado pela ECU
-		osEventFlagsWait(ECU_control_event_id, ALL_ERRORS_FLAG, osFlagsWaitAny | osFlagsNoClear, osWaitForever);
+		osEventFlagsWait(ECU_control_event_id, ALL_SEVERE_ERROR_FLAG, osFlagsWaitAny | osFlagsNoClear, osWaitForever);
 		uint32_t error_flags = osEventFlagsGet(ECU_control_event_id);
-		error_flags = error_flags & ALL_ERRORS_FLAG; //filtra apenas flags de erros, ignorando as outras
+		error_flags = error_flags & ALL_SEVERE_ERROR_FLAG; //filtra apenas flags de erros severos, que precisam sair de RTD, ignorando as outras
 		switch (error_flags) {
-			case APPS_ERROR_FLAG:
-				//TODO: tratar erro de APPS
-				break;
 			case INVERTER_COMM_ERROR_FLAG:
-				//TODO: tratar erro de comunicacao com o inversor
+				exit_RTD();
 				break;
 		}
```

Comment:
repara que ele vai chegar no final desse bloco e vai voltar lá pro início do for, e aí ele vai esperar apertar o botão de RTD de novo (e nesse momento a tarefa vai entrar em deadlock, pq n faz sentido apertar o RTD nesse estado. Vc precisa bolar uma forma dele repetir só esse trecho de código até que saia do modo de RTD. Quando sair, aí vc volta pro início da tarefa

Reply:
Minha ideia era essa msm.  Ele vai sair de RTD e voltar pra esperar apertar o botão. Caso o erro tenha desaparecido o piloto consegue acionar RTD. Caso o erro mantenha o RTD fica bloqueado. Ai pensei em fazer verificação de comunicação e caso esteja tudo certo ele sempre da clearflag pra limpar se tiver tido erro antes.